ddtrace/contrib/aiopg/connection.py

```diff
@@ -0,0 +1,92 @@
+import asyncio
+
+import wrapt
+from aiopg.utils import _ContextManager
+from .. import dbapi
+from ddtrace import Pin
+from ext import sql
+from aiopg.cursor import Cursor
+
+
+class AIOTracedCursor(wrapt.ObjectProxy):
+    """ TracedCursor wraps a psql cursor and traces it's queries. """
+
+    def __init__(self, cursor, pin):
+        super(AIOTracedCursor, self).__init__(cursor)
+        pin.onto(self)
+        name = pin.app or 'sql'
+        self._datadog_name = '%s.query' % name
+
+    @asyncio.coroutine
+    def _trace_method(self, method, resource, extra_tags, *args, **kwargs):
+        pin = Pin.get_from(self)
+        if not pin or not pin.enabled():
+            result = yield from method(*args, **kwargs)  # noqa: E999
+            return result
+        service = pin.service
+
+        with pin.tracer.trace(self._datadog_name, service=service,
+                              resource=resource) as s:
+            s.span_type = sql.TYPE
+            s.set_tag(sql.QUERY, resource)
+            s.set_tags(pin.tags)
+            s.set_tags(extra_tags)
+
+            try:
+                result = yield from method(*args, **kwargs)
+                return result
+            finally:
+                s.set_metric("db.rowcount", self.rowcount)
+
+    @asyncio.coroutine
+    def executemany(self, query, *args, **kwargs):
+        # FIXME[matt] properly handle kwargs here. arg names can be different
+        # with different libs.
+        result = yield from self._trace_method(
+            self.__wrapped__.executemany, query, {'sql.executemany': 'true'},
+            query, *args, **kwargs)  # noqa: E999
+        return result
+
+    @asyncio.coroutine
+    def execute(self, query, *args, **kwargs):
+        result = yield from self._trace_method(
+            self.__wrapped__.execute, query, {}, query, *args, **kwargs)
+        return result
+
+    @asyncio.coroutine
+    def callproc(self, proc, args):
+        result = yield from self._trace_method(
+            self.__wrapped__.callproc, proc, {}, proc, args)  # noqa: E999
+        return result
+
+    # aiopg doesn't support __enter__/__exit__ yet so we're adding it here to
+    # support unittests with both styles
+    if getattr(Cursor, '__enter__', None) is None:
```

Comment:
my concern is that in the current `aiopg` version it's not available (as I read in your comment) so with this patch what's happening is:
1) `aiopg` doesn't expose a context manager
2) people use our instrumented cursor
3) magically they have a context manager (because that will be evaluated to `None`)
4) they will start using a context manager thinking that it's built-in in the library
5) they remove the instrumentation / move to another vendor
6) the application crashes badly :(

What I'm thinking is:
* can't we simply ignore that behavior and don't test it, since it's not available in the library? We will add it in a later version when it will be available
* if we really need it for some syntactic sugar in the tests, can't we simply move that in a Mixin that is used only in the test itself?

I still prefer not covering cases that are not exposed by the library because it could be too much error prone.